src/pages/Validator/AutocompleteList.js

```diff
@@ -2,36 +2,78 @@ import React from "react";
 
 export default function AutocompleteList({
   entities = {},
+  showAttributes,
   inputVal,
   onSelect,
   onCreate,
 }) {
-  // Calls onSelect to insert token into the UI
+  /*
+   * Inserts token into the UI, replacing whatever text the user already wrote 
+   * in the token
+   */
   let handleSelect = (entity) => {
     onSelect(entity);
   };
 
-  let entitiesSeen = [];
-
-  // Create a list of synonyms, filter them with the input value, and then return a token if the entity isn't already displayed
-  let shownOptions = Object.keys(entities)
-    .filter((entity) => entity.toUpperCase().startsWith(inputVal.toUpperCase()))
-    .map((entity) => {
-      if (!entitiesSeen.includes(entities[entity])) {
-        entitiesSeen.push(entities[entity])
-        return (
-          <li key={entity} onClick={handleSelect.bind(this, entities[entity])} >
-            {entities[entity]}
-          </li >
-        )
+  
+  /*
+   * Returns a list of synonym autocomplete options, filtered by the input
+   */
+  let listSynonyms = () => {
+    let entitiesSeen = [];
+    // Map entities and their synonyms to their correct entity
+    let synonyms = {};
+    for (let entity in entities) {
+      synonyms[entity] = entity;
+      for (let syn of entities[entity]['synonyms']) {
+        synonyms[syn] = entity;
       }
+    }
+    return (
+      Object.keys(synonyms)
+      .filter(entity => entity.toUpperCase().startsWith(inputVal.toUpperCase()))
+      .map((entity) => {
+        if (!entitiesSeen.includes(synonyms[entity])) {
+          entitiesSeen.push(synonyms[entity])
```

Comment:
We can make this O(1) by making `entitiesSeen` a Set

Reply:
🧠 👍